extensions/terminal-suggest/src/terminalSuggestMain.ts

```diff
@@ -0,0 +1,248 @@
+
+let cachedAvailableCommands: Set<string> | undefined;
+let cachedBuiltinCommands: Map<string, string[]> | undefined;
+
+function getBuiltinCommands(shell: string): string[] | undefined {
+	try {
+		const shellType = path.basename(shell);
+		const cachedCommands = cachedBuiltinCommands?.get(shellType);
+		if (cachedCommands) {
+			return cachedCommands;
+		}
+		const options: ExecOptionsWithStringEncoding = { encoding: 'utf-8', shell };
+		switch (shellType) {
+			case 'bash': {
+				const bashOutput = execSync('compgen -b', options);
+				const bashResult = bashOutput.split('\n').filter(cmd => cmd);
+				if (bashResult.length) {
+					cachedBuiltinCommands?.set(shellType, bashResult);
+					return bashResult;
+				}
+				break;
+			}
+			case 'zsh': {
+				const zshOutput = execSync('printf "%s\\n" ${(k)builtins}', options);
+				const zshResult = zshOutput.split('\n').filter(cmd => cmd);
+				if (zshResult.length) {
+					cachedBuiltinCommands?.set(shellType, zshResult);
+					return zshResult;
+				}
+			}
+			case 'fish': {
+				// TODO: ghost text in the command line prevents
+				// completions from working ATM for fish
+				const fishOutput = execSync('functions -n', options);
+				const fishResult = fishOutput.split(', ').filter(cmd => cmd);
+				if (fishResult.length) {
+					cachedBuiltinCommands?.set(shellType, fishResult);
+					return fishResult;
+				}
+				break;
+			}
+		}
+		// native pwsh completions are builtin to vscode
+		return;
+
+	} catch (error) {
+		console.error('Error fetching builtin commands:', error);
+		return;
+	}
+}
+
+
+export async function activate(context: vscode.ExtensionContext) {
+	context.subscriptions.push(vscode.window.registerTerminalCompletionProvider({
+		id: 'terminal-suggest',
+		async provideTerminalCompletions(terminal: vscode.Terminal, terminalContext: { commandLine: string; cursorPosition: number }, token: vscode.CancellationToken): Promise<vscode.TerminalCompletionItem[] | undefined> {
+			if (token.isCancellationRequested) {
+				return;
+			}
+
+			const availableCommands = await getCommandsInPath();
+			if (!availableCommands) {
+				return;
+			}
+
+			// TODO: Leverage shellType when available https://github.com/microsoft/vscode/issues/230165
+			const shellPath = 'shellPath' in terminal.creationOptions ? terminal.creationOptions.shellPath : vscode.env.shell;
+			if (!shellPath) {
+				return;
+			}
+
+			const builtinCommands = getBuiltinCommands(shellPath);
+			builtinCommands?.forEach(command => availableCommands.add(command));
+
+			const prefix = getPrefix(terminalContext.commandLine, terminalContext.cursorPosition);
+			let result: vscode.TerminalCompletionItem[] = [];
+			const specs = [codeCompletionSpec, codeInsidersCompletionSpec];
+			for (const spec of specs) {
+				const specName = getLabel(spec);
+				if (!specName || !availableCommands.has(specName)) {
+					continue;
+				}
+				if (terminalContext.commandLine.startsWith(specName)) {
```

Comment:
this condition means that the completion won't work if a user does `command1 && command2` and `command2` is `code` or `code-insiders` for example. 

It will work if the spec is `command1` though. 

I think we can defer that.